src/metabase/notification/send.clj

```diff
@@ -62,12 +63,17 @@
                               (try
                                 (channel/send! channel message)
                                 (catch Exception e
-                                  (when (should-retry-sending? e (:type channel))
+                                  (let [skip-retry? (should-skip-retry? e (:type channel))
+                                        new-e       (ex-info (ex-message e)
+                                                             (assoc (ex-data e) ::skip-retry? skip-retry?))]
                                     (vswap! retry-errors conj {:message   (u/strip-error e)
                                                                :timestamp (t/offset-date-time)})
-                                    (log/warnf e "[Notification %d] Failed to send to channel %s , retrying..."
-                                               notification-id (handler->channel-name handler))
-                                    (throw e)))))
```

Comment:
`task-history/with-task-history` will record a task to be failed when the the body throw an exception. And in here we currently skip throwing the exception if `should-retry-sending?` returns false.

The fix is to make sure we throw an exception in all cases but still kip the skip retry behavior. We do this by defining a custom `retry-on-exception-pred` for the retry config.